In [10]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn import datasets
from sklearn.linear_model import LogisticRegression as skLogisticRegression
from sklearn.metrics import accuracy_score


In [9]:
def sigmoid (x):
    return (1/(1+np.exp(-x)))

In [19]:
# Model implementation
class LogisticRegression:

    def __init__(self,alpha = 0.001, n_iters = 1000):
        self.alpha = alpha
        self.n_iters = n_iters
        self.weights = None
        self.bias = None


    def fit(self,X,y):
        m , n = X.shape 
        self.weights = np.zeros(n)
        self.bias = 0

        for _ in range(self.n_iters):
            # model
            
            linear_pred = np.dot(self.weights,X.T) +self.bias
            y_pred = sigmoid(linear_pred)
            # gradients 
            dj_dw = (1/m) * np.dot((y_pred- y),X)
            dj_db = (1/m) * np.sum(y_pred- y)

            self.weights -= self.alpha * dj_dw
            self.bias -= self.bias * dj_db


    def predict(self, X):
        linear_pred = np.dot(self.weights,X.T) +self.bias
        y_pred = sigmoid(linear_pred)
        class_pred = [0 if y<0.5 else 1 for y in y_pred]
        return class_pred
    

    def accuracy(self,y_test , y_pred):
        return np.sum(y_pred == y_test)/len(y_test)


In [20]:
# Model Testing
bc = datasets.load_breast_cancer()
X, y = bc.data ,bc.target
X_train , X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=1234)


clf = LogisticRegression(alpha=0.01)
clf.fit(X_train,y_train)
pred = clf.predict(X_test)
acc = clf.accuracy(y_test,pred)

print(f"accuracy = {acc}")


C:\Users\mahamedhat\AppData\Local\Temp\ipykernel_37520\1469425069.py:2: RuntimeWarning: overflow encountered in exp
  return (1/(1+np.exp(-x)))


accuracy = 0.9210526315789473


In [21]:
#sklearn model
model = skLogisticRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

print(f"accuracy = {accuracy}")

accuracy = 0.9473684210526315


c:\Users\mahamedhat\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## Proof of concept
My model achieved an accuracy of 0.921, while the scikit-learn model achieved an accuracy of 0.947. The difference in accuracy between my model and the scikit-learn model suggests that my model is slightly less accurate than the industry standard. However, an accuracy of over 90% indicates that my model is still performing well and may be suitable for certain applications. It's important to note that the accuracy of my model may vary depending on the specific data it is trained on and the features used. Further experimentation and testing may be necessary to determine the true accuracy of my model in different contexts.